src/huggingface_hub/inference_api.py

```diff
@@ -0,0 +1,108 @@
+from .validation import ALL_TASKS, check_inputs, check_params
+
+
+logger = logging.getLogger(__name__)
+
+
+ENDPOINT = "https://api-inference.huggingface.co/pipeline"
+
+
+class InferenceApi:
+    """Client to configure requests and make calls to the HuggingFace Inference API.
+
+    Example:
+
+            >>> from huggingface_hub.inference_api import InferenceApi
+
+            >>> # Mask-fill example
+            >>> api = InferenceApi("bert-base-uncased")
+            >>> api.set_inputs(inputs="The goal of life is [MASK].")
+            >>> api.call()
+
+            >>> # Question Answering example
+            >>> api = InferenceApi("deepset/roberta-base-squad2")
+            >>> api.set_inputs(question="What's my name?", context="My name is Clara and I live in Berkeley.")
+            >>> api.call()
+
+            >>> # Zero-shot example
+            >>> api = InferenceApi("typeform/distilbert-base-uncased-mnli")
+            >>> api.set_inputs(inputs="Hi, I recently bought a device from your company but it is not working as advertised and I would like to get reimbursed!")
+            >>> api.set_params(candidate_labels=["refund", "legal", "faq"])
+            >>> api.call()
+    """
+
+    def __init__(
+        self, repoId: str, task: Optional[str] = None, gpu: Optional[bool] = False
+    ):
+        """Inits InferenceApi headers and API call information.
+
+        Args:
+            repoId (``str``): Id of model (e.g. `bert-base-uncased`).
+            task (``str``, `optional`, defaults ``None``): Whether to force a task instead of using task specified in repository.
+            gpu (``bool``, `optional`, defaults ``None``): Whether to use GPU instead of CPU for inference(requires Startup plan at least).
+        """
+        token = HfFolder.get_token()
+        if token is None:
+            raise EnvironmentError("A Hugging Face token was not found.")
+
+        # Configure task
+        modelInfo = HfApi().model_info(repo_id=repoId, token=token)
+        if not modelInfo.pipeline_tag and not task:
+            raise ValueError(
+                "Task not specified in the repository. Please add it to the model card using pipeline_tag (https://huggingface.co/docs#how-is-a-models-type-of-inference-api-and-widget-determined)"
+            )
+
+        if task and task != modelInfo.pipeline_tag:
+            if task not in ALL_TASKS:
+                raise ValueError(f"Invalid task {task}. Make sure it's valid.")
+
+            logger.warning(
+                "You're using a different task than the one specified in the repository. Be sure to know what you're doing :)"
+            )
+            self.task = task
+        else:
+            self.task = modelInfo.pipeline_tag
+
+        # Configure url, headers and options
+        self.api_url = f"{ENDPOINT}/{self.task}/{repoId}"
+        self.headers = {"authorization": "Bearer {}".format(token)}
+        self.options = {"wait_for_model": True, "use_gpu": gpu}
```

Comment:
I am not 100% convinced about waiting for model by default.

We should either load it right here. or like the widget try without `wait_for_model` and retry with waiting.

Reply:
What is the advantage of not waiting for the model by default? Do we expect our users to make concurrent requests?

Reply:
Personally I think we should `wait_for_model` by default but display a info log that the model is being loaded (when not already loaded) (based on a first HTTP response header being streamed, or something like that)

I know there was a technical constraint to not do it like that in the widgets, but we should maybe revisit if we can.

Reply:
> What is the advantage of not waiting for the model?
You can display a nice error message / loading bar. That's about it.

Having some system hanging for no valid reason is a great way to get users to quit. That's why IMO wait_for_model is opt-in (at least you know what you are waiting for).

We could also simply query the endpoint to know if it's loaded first or not `https://api-inference.huggingface.co/status/{MODEL_NAME}`. It does not currently give a loading time estimate, but it's easy to add.

> Personally I think we should wait_for_model by default but display a info log that the model is being loaded (when not already loaded) (based on a first HTTP response header being streamed, or something like that)
> 
> I know there was a technical constraint to not do it like that in the widgets, but we should maybe revisit if we can

I don't think we can start sending HTTP headers before having the content-length, can we ?

Visiting some API that keeps persistent connection (websocket, gRPC, ...) is something of interest, but probably out of scope here.

Reply:
> I don't think we can start sending HTTP headers before having the content-length, can we ?

I think you can

Reply:
You are correct, HTTP 1.1 enables chunked answers ofc.
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html

However I feel like it's quite a stretch from the HTTP spec. I quickly checked Starlette, and there is no easy mechanism to use that, I can't even imagine the client libraries being able to use that behavior simply either. 2 HTTP calls seems better/simpler IMO, or switching completely to some sort of persistent connection mechanism or maybe some sort of queue mechanism.

I do agree that it is currently a bit of friction when you have to wait for the model, and changing the actual call to start waiting for the answer is not great. Anything to help with that would be good.

Reply:
i suspect that using an async-first (or event centric) library like https://docs.aiohttp.org/en/stable/ would get us there pretty easily (I mean, it's trivial in JS 😎 ), but I might be overseeing some complexities.

Same on the server side: there's asynchronous in the name of https://asgi.readthedocs.io/en/latest/

on the server-side I've used `Content-Type: 'text/event-stream'` in the past, it's very well supported

Maybe let's get some feedback/intuition from a few other team members on this. @n1t0 @Pierrci @SBrandeis maybe?

Reply:
pinging again @n1t0 @Pierrci @SBrandeis on this point

Reply:
In a fully async world, I would probably expect to get a response once the model is loaded, waiting for it if needed. But here in Python, I must say I don't know which one is best. Even though I feel like waiting for the model to load might be the most intuitive, in some cases, it might be a bad experience.
What about making it an argument of `__call__` to let the user chose whether they want to wait?

@Narsil can you show us what you expect the user to be doing (ie the code they would have to write) if it were to return immediately?

Reply:
Oww, I wasn't suggesting that the function returns immediately, but that we were able to prompt something to the user. prompt-wise:

```python
myapi("My prompt")
>>> WARNING: model is not loaded waiting for it to load [|||||||||||||............] (24s / 60)
>>> (Hang, and end with result) "My prompt is something super nice"
```
What I was suggesting is to do sub sequential calls so that users are aware about the loading times (because it's almost always quite long of the orders of tens of seconds compared to the inference.
Just like the widgets in fact, call once without waiting, have some kind of loading being displayed, then launch while waiting so you get the result as soon as possible

Reply:
I like this idea. I will work in a follow-up PR to integrate this plus a couple of other new ideas :)